File: C++/1-D Dynamic Programming/house_robber.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        /*
            Problem: House Robber
            - Given an array `nums` where each element represents the amount of money at each house.
            - You cannot rob two adjacent houses.
            - Find the maximum amount of money you can rob.

            Approach:
            - Dynamic Programming (space-optimized).
            - Use two variables:
              - rob1: max money if we rob up to house i-2
              - rob2: max money if we rob up to house i-1
            - At each house, decide:
              - Rob this house → add current house's money to rob1
              - Or skip this house → keep rob2
            - The maximum of these two becomes the new `rob2`.

            Time Complexity: O(n)
                - We iterate through the array once.
            Space Complexity: O(1)
                - We only use two variables to keep track of the state.
        */

        int rob1 = 0, rob2 = 0;

        for (int n : nums) {
            // temp stores the maximum money after deciding whether to rob this house or not
            int temp = max(n + rob1, rob2);
            rob1 = rob2;  // Move rob2 to rob1 for the next iteration
            rob2 = temp;  // Update rob2 to the new maximum
        }

        return rob2; // rob2 holds the maximum money that can be robbed
    }
};

/*
================ LINE-BY-LINE SYNTAX EXPLANATION =================

#include <bits/stdc++.h>
    - Includes all standard C++ libraries like vector, iostream, etc.

using namespace std;
    - Allows using standard library names directly without the `std::` prefix.

class Solution {
public:
    - Defines a class named Solution with public member functions.

int rob(vector<int>& nums)
    - Function named `rob` that takes a reference to a vector of integers `nums` and returns an integer.

int rob1 = 0, rob2 = 0;
    - Initialize two variables to track maximum money if you rob up to house i-2 and i-1.

for (int n : nums)
    - Range-based for loop to iterate through each element of `nums`. Here `n` is each house's money.

int temp = max(n + rob1, rob2);
    - Decide to rob the current house (n + rob1) or skip it (rob2), and take the maximum.

rob1 = rob2;
    - Update rob1 for the next iteration to be rob2.

rob2 = temp;
    - Update rob2 to the new maximum after this house.

return rob2;
    - Return the maximum money that can be robbed.

*/

File: python/1-D Dynamic Programming/house_robber.py
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        # rob1: max money that can be robbed up to the house before the previous one
        # rob2: max money that can be robbed up to the previous house
        rob1, rob2 = 0, 0

        # Iterate through each house's money
        for n in nums:
            # For each house, decide to either:
            # 1. Rob it (add current house's money to rob1, which is the max up to house i-2)
            # 2. Skip it (keep rob2, which is the max up to house i-1)
            temp = max(n + rob1, rob2)
            rob1 = rob2  # Move rob2 to rob1 for next iteration
            rob2 = temp  # Update rob2 to the new max
        return rob2  # rob2 contains the max money that can be robbed

# Time Complexity: O(n), where n is the number of houses (nums)
# Space Complexity: O(1), since only two variables are used regardless of input size

class Solution2:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        
        if len(nums) == 2:
            return max(nums[0], nums[1])

        dp = [0] * len(nums)
        dp[0] = nums[0]
        dp[1] = max(nums[1], nums[0])
        n = len(nums)

        for i in range(2, n):
            dp[i] = max(nums[i] + dp[i - 2], dp[i - 1])
        
        return dp[n - 1]

# -------------------
# Test Cases
# -------------------
def test_house_robber():
    s = Solution()
    # Test case 1: Example from Leetcode
    assert s.rob([1,2,3,1]) == 4  # Rob house 1 and 3 (1 + 3)
    # Test case 2: Another example
    assert s.rob([2,7,9,3,1]) == 12  # Rob house 1, 3, 5 (2 + 9 + 1)
    # Test case 3: Single house
    assert s.rob([5]) == 5
    # Test case 4: Two houses
    assert s.rob([2,1]) == 2
    # Test case 5: All houses have same value
    assert s.rob([2,2,2,2]) == 4
    # Test case 6: Empty list
    assert s.rob([]) == 0
    print("All test cases passed!")

if __name__ == "__main__":
    test_house_robber()
